game-core/src/main/java/games/strategy/engine/data/MutableProperty.java

```diff
@@ -145,6 +146,20 @@ public void resetValue() {
     return ofString(stringSetter, noGetter(), noResetter());
   }
 
+  public static <T> MutableProperty<T> ofMapper(
+      final ThrowingFunction<String, T, Exception> mapper,
+      final ThrowingConsumer<T, Exception> setter,
+      final Supplier<T> getter,
+      final Supplier<T> defaultValue) {
+    return new MutableProperty<>(setter, o -> setter.accept(mapper.apply(o)), getter, () -> {
+      try {
+        setter.accept(defaultValue.get());
+      } catch (final Exception e) {
+        throw new IllegalStateException("Unexpected Error while resetting value", e);
```

Comment:
Not sure if it's more appropriate to throw a checked or unchecked exception here.  It was an easy decision in `setValue()` because the caller is providing (possibly invalid) input.  Here, if an error occurs during a reset, it would seem to be a programmer error, since it's all based on the internal implementation of the target.

Since we normally only allow a resetter to throw unchecked exceptions, we're basically saying here, "if your setter or mapper throw during a reset, it's your fault, not the caller's."  I guess I can live with that. :smile: